### DISCUSSION TITLE
How to use `%>>%` / perform side effects

### DISCUSSION THREAD

**Discussion by yogat3ch** [r-lib/mirai#72] [Q&A] [answered]
Hi `mirai` devs
I'm unable to make sense of how to feasibly put `mirai` to use from the examples provided in the documentation. Below is a simple use case to demonstrate.
I would expect `env$test` to evaluate to `TRUE`, but it's just `NULL`
```r
library(mirai)
env <- new.env()
mirai::.(
  mirai::mirai({
    Sys.sleep(5)
    TRUE
  }) %>>%
  assign(x = "test", envir = env)
 )
Sys.sleep(6)
env$test
``` 

I've pored over the documentation for far too long now and just can't make sense of how to use `mirai` in production, any help?

**Comment by shikokuchuo** [comment, marked as answer]
The deferred execution pipe `%>>%` requires the `mirai` to be assigned to a variable outside of the `.()` scope and a check for resolution by `unresolved()` or directly accessing the active binding `b$data`.

So for you example above:

```r
library(mirai)
env <- new.env()
m <- mirai({
  Sys.sleep(5)
  TRUE
})
b <- .(m %>>% assign(x = "test", envir = env))
Sys.sleep(6)
if (!unresolved(b)) env$test
```
However if you only care about the side effect (performing the assignment to the environment) rather than using the return value, then using mirai promises and a different pipe may be a better fit. For example:

```r
library(mirai)
library(promises)
library(mirai.promises)
env <- new.env()
mirai({
  Sys.sleep(5)
  TRUE
}) %...>% assign(x = "test", envir = env)
Sys.sleep(6)
env$test
```

**Reply by yogat3ch** [reply]
The second example with `mirai.promises` is the ticket, that's exactly what I'm looking for. It seems like `mirai.promises` is definitely what I'm interested in, thank you @shikokuchuo !
